test.sh

```diff
@@ -0,0 +1,455 @@
+#==========================================
+
+printf "DDlog test suite\n\n"
+testsuite_setup
+
+if ( [ "$#" -eq 0 ] || [ "x$1" == "xhelp" ] || [ "x$1" == "x--help" ] ); then
+    printf "Usage: ${0} test_or_test_group1 test_or_test_group2 ...\n"
+    printf "   or: ${0} all\n\n"
+    printf "Available test groups:\n"
+    for group_with_descr in "${test_groups[@]}"
+    do
+        IFS=":" read -ra tokens <<< "$group_with_descr"
+        group=${tokens[0]}
+        description=${tokens[1]}
+
+        printf "\n    ${group}: ${description}\n"
+
+        eval "tests=(\"\${${group}[@]}\")"
+        for tst_with_descr in "${tests[@]}"
+        do
+            IFS=":" read -ra tokens <<< "$tst_with_descr"
+            tst=${tokens[0]}
+            description=${tokens[1]}
+            printf "        %-25s %s\n" "${tst}" "${description}"
+        done
+    done
+else
+    if [ "x$1" == "xall" ]; then
+        test_list=()
+        for group_with_descr in "${test_groups[@]}"
+        do
+            IFS=":" read -ra tokens <<< "$group_with_descr"
+            group=${tokens[0]}
+            test_list+=(${group})
+        done
+    else
+        test_list="$@"
+    fi
+
+    # Make a list of tests to run.
+    all_tests=()
+
+    for tst in "${test_list[@]}"
+    do
+        if [[ "${test_groups[@]}" =~ "${tst}" ]]; then
+            #echo "test group '${tst}'"
+            eval "tests=(\"\${${tst}[@]}\")"
+            for tst_with_descr in "${tests[@]}"
+            do
+                IFS=":" read -ra tokens <<< "$tst_with_descr"
+                tst_func=${tokens[0]}
+
+                # Only add test if not already in the list.
+                if ! [[ " ${all_tests[@]} " =~ " ${tst_func} " ]]; then
+                    all_tests+=(${tst_func})
+                fi
+            done
+
+        else
+            # Only add test if not already in the list.
+            if ! [[ " ${all_tests[@]} " =~ " ${tst} " ]]; then
+                all_tests+=(${tst})
+            fi
+        fi
+    done
+
+    echo "Running the following tests: ${all_tests[@]}"
+    echo ""
+
+    # Validate the resulting list of tests.
+    for tst in "${all_tests[@]}"; do
+        if ! ([ -n "$(type -t ${tst})" ] && [ "$(type -t ${tst})" = function ]); then
+            echo "Unknown test '${tst}'"
+            fail=1
+        fi
+    done
+
+    if [ " ${fail} " == " 1 " ]; then
+        printf "${RED}FAIL${NC}\n"
+        exit 1
+    fi
+
+    # Run the tests.
+    passed=0
+    failed=0
+    for tst in "${all_tests[@]}"; do
+        printf "%-25s %s" "${tst}"
+        start=`date +%s`
+        if run_test "${tst}"; then
+            ok
+            passed=$((passed+1))
+        else
+            fail
+            if [ -z "${NO_STD_REDIRECT}" ]; then
+                printf " [output saved in '${log_file_name}']"
+            fi
+            failed=$((failed+1))
+        fi
+        end=`date +%s`
+        runtime=$((end-start))
+        printf " (${runtime}s)\n"
+    done
```

Comment:
For scripts of this complexity I usually use Python, but hopefully this won't change much from now on.